removes contaminant sequences from representative sequences

jordenrabasco opened the pull request:
This pull request adds additional functionality by allowing users to remove contaminant identified sequences from the representative sequences object created from q2-DADA2

jordenrabasco committed:
initial integration

-working integration of decontam and quality control

jordenrabasco committed:
clean plugin_steup.py

-removes redundant code and updates look of code

jordenrabasco committed:
Clean setup.py

-improves spacing

jordenrabasco committed:
clean transfomer.py

-removes old code and updates spacing

jordenrabasco committed:
Clean test_stats.py

-updates spacing

jordenrabasco committed:
Cleans test_q2_quality_control

-removes redundant code and cleans up spacing

jordenrabasco committed:
Cleans _stats.py and quality_control.py

-removes old code and updates spacing

jordenrabasco committed:
rounding error in histogram

-fixes rounding error in mod function in the histogram

jordenrabasco committed:
clean run_decontam.R

-removes local testing paths form file

jordenrabasco committed:
Update run_decontam.R

-removes spacing and extra comments

jordenrabasco committed:
Update error message

-updates R error messages

jordenrabasco committed:
Update citations.bib

-updates citations with decontam paper

jordenrabasco committed:
Update plugin_setup.py

-updates decontam remove description

jordenrabasco committed:
lint update: _stats, 1

-lint update

jordenrabasco committed:
lint update: _stats.py 2, _vizualizer.py 1

-lint update

jordenrabasco committed:
lint update: _vizualizer.py 2

-lint update

jordenrabasco committed:
lint update: _vizulaizer.py 3, _transformer.py 1

-lint update

jordenrabasco committed:
lint update: plugin_setup.py 1

-lint update

jordenrabasco committed:
lint update: plugin_setup.py 2

-lint update

jordenrabasco committed:
lint update: plugin_setup.py 3, quality_control.py 1

-lint update

jordenrabasco committed:
lint update: quality_control.py 2, test_quality_control.py 1

-lint update

jordenrabasco committed:
lint update: quality_control.py 3, test_quality_control.py 2

-lint update

jordenrabasco committed:
lint update: test_quality_control.py 3

-lint update

jordenrabasco committed:
lint update: test_quality_control.py 4

lint update

jordenrabasco committed:
lint update: test_quality_control.py 5, test_stats.py 1

-lint update

jordenrabasco committed:
lint update: test_stats.py 2

-lint update

jordenrabasco committed:
lint update: test_stats.py 3

-lint update

jordenrabasco committed:
lint update: setup.py 1

-lint update

jordenrabasco committed:
updates yaml file and graph

- adds R package to yaml file
- removes decrepit reference from graphing function

jordenrabasco committed:
testing error potential fix 1

-error testing in build fix

jordenrabasco committed:
testing error potential fix 2

jordenrabasco committed:
testing error potential fix 3

-move decontam tests

jordenrabasco committed:
testing error potential fix 4

jordenrabasco committed:
testing error potential fix 5

jordenrabasco committed:
test recongifuration

-reconfigures tests to come into line with CI standards

jordenrabasco committed:
lint update

jordenrabasco committed:
lint update

jordenrabasco committed:
lint update

jordenrabasco committed:
update file discovery for ci

jordenrabasco committed:
refactor and file removal

-removal of decrepit files and refactorization
-added line in setup.py to make run_decontam.R discoverable

jordenrabasco committed:
User side requested changes

-updates user api to come into line with desired changes

jordenrabasco committed:
Transpose issue fix

-fixes implementation issue associated with transposition of the ASV input matrix.
- direct fix to issue referenced in comment:
https://github.com/qiime2/q2-quality-control/pull/74#discussion_r1112298912

jordenrabasco committed:
Implementation update

-resolves all implementation comments

jordenrabasco committed:
implementation lint fix

-lint 1

jordenrabasco committed:
implementation lint 2

-lint fix 2

jordenrabasco committed:
implementation lint 3

-lint fix 3

jordenrabasco committed:
tranpose and desc format changes

-requested changes fix

jordenrabasco committed:
Control Samples error catching

-Makes sure that there are an appropriate amount of Control samples in both the ASV and metadata table

jordenrabasco committed:
removes extraneious comments

jordenrabasco committed:
clean up 1

jordenrabasco committed:
drop column update

-updates what columns to drop by being explicit and updates tests

jordenrabasco committed:
update histogram to remove NAs

-removes NA counts from histogram

jordenrabasco committed:
lint update histogram

-fixes minor lint errors

ebolyen committed:
BUG: fix ID matching

ebolyen committed:
lint

jordenrabasco committed:
tests_added_bug_fix

-changes line in run_decontam.R
-adds tests to test_decontam.py

jordenrabasco committed:
lint

jordenrabasco committed:
catch included for no metadata for samples in ASV table

-error value added

jordenrabasco committed:
Merge branch 'qiime2:master' into master

jordenrabasco committed:
Update run_decontam.R

- Updates run_decontam.R  to allow for special characters in the metadata headings

jordenrabasco committed:
initial push - hierarchical pipeline

-untested base functionality with no errors

jordenrabasco committed:
alogirthm update

jordenrabasco committed:
decontam_fast update

-intial commit fix

jordenrabasco committed:
metadata update error

-initial fix to metadata updating

jordenrabasco committed:
Update nomenclature

-updates file names
-deletes print statments

jordenrabasco committed:
optimized table generation

-Optimizes split table generation
-remove print statments

jordenrabasco committed:
further cleanup

-cleans up code

jordenrabasco committed:
initial push of updated visualizer

-initial push

jordenrabasco committed:
line update

-update when to use the line

jordenrabasco committed:
vizualizer fix

-fixes rounding error in vizualizer coloring

jordenrabasco committed:
wip

jordenrabasco committed:
Pipe update

-updated how data was being transfered

jordenrabasco committed:
lint fix _pipelines.py

jordenrabasco committed:
lint vizualizer.py

jordenrabasco committed:
lint update

jordenrabasco committed:
viz coloring fix

-fixes viz coloring with rounding issues in python

jordenrabasco committed:
lint fix

ebolyen committed:
LINT: remove parsl scraps

ebolyen committed:
lint: workspace file

ebolyen committed:
DEP: add feature-table dependency

jordenrabasco committed:
rep_seq_inital function push

-includes functionality related to the rep_seqs_trimming

jordenrabasco committed:
test first attempt

-test for rep_seqs

jordenrabasco committed:
lint updatea

jordenrabasco committed:
testing_update

- test update

jordenrabasco committed:
removal bad files

jordenrabasco committed:
Merge branch 'dev' into bad_seq_removal_expansion

jordenrabasco committed:
lint 1

linting

jordenrabasco committed:
Merge branch 'bad_seq_removal_expansion' of https://github.com/jordenrabasco/q2-quality-control into bad_seq_removal_expansion

jordenrabasco committed:
lint 2

linting

jordenrabasco committed:
lint 3

jordenrabasco committed:
minor bug error

jordenrabasco committed:
update_testing_data

-updates testing data

jordenrabasco committed:
one line change

gregcaporaso commented on q2_quality_control/decontam.py:
```diff
@@ -165,22 +168,35 @@ def decontam_identify(table: pd.DataFrame,
         return _decontam_identify_helper(track_fp, method)
 
 
+def _filepath_to_sample_single(fp):
```
This doesn't ever seem to be used, but if I'm wrong about that let's talk about better ways to get this information. This can be problematic (e.g., a sample id is allowed to have an `_` in it).

gregcaporaso commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -9,6 +9,7 @@
 import pandas as pd
 import qiime2
 import biom
+import skbio
```
This would be a new dependency for this plugin, and I think it should be possible to do without it, so we should avoid it to keep the dependency stack as minimal as possible.

gregcaporaso commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -121,26 +126,58 @@ def setUp(self):
+            qiime2.Metadata).to_dataframe()
+        output_table = qiime2.Artifact.load(
+            self.get_data_path('expected/output_remove_contam_table.qza'))
+        self.rep_seq_output_table = output_table.view(
```
It's not clear why this object would be called `rep_seq_output_table`. The `rep_seq` bit makes it really confusing, and `output` is ambiguous here. Generally speaking, try to think about making your variable names self-documenting. That makes it easier for others to follow, so helps with review and maintenance (by you or others) in the future easier. 

In this case, that could include calling the sequence-related variables `sequences` and the table-related variables `table` (for example). I also recommend using `observed` (or `actual`) and `expected`, rather than `output` for clarity. 

This code would be much easier to follow if you were comparing `observed_sequences` to `expected_sequences` and `observed_table` to `expected_table`.

gregcaporaso replied:
You also don't need to add these as attributes on the class (i.e., start them with `self`) if you're only using them in one function.

gregcaporaso commented on q2_quality_control/tests/test_decontam.py:
```diff
@@ -121,26 +126,58 @@ def setUp(self):
+            rep_seqs=self.rep_seq_removal_table)
+        temp_table = output_asv_table.to_dataframe().transpose()
+        temp_table.index.name = 'id'
+        temp_table.columns = range(1, len(temp_table.columns) + 1)
```
It looks like you're throwing away the column names here. Is that intended? I am doing some testing with this and running into an issue where the column names aren't aligned, when they should be.

gregcaporaso replied:
I think the reason that your ids may not be lining up is because the input and output feature tables share no features. I'm submitting some edits that I made where I broke your .qza files out for easier access in tests, and when I do the following:

```bash
(q2dev-amplicon) ➜  expected git:(gregs-review-pr81) ✗ biom table-ids --observations -i remove_contam_test_table/feature-table.biom | sort | head
01173073677a287321be3484fbed0007
0160e14a78b18b903618f11bc732746e
01b99cb344ed2530f7d80897ffe257a9
01ce91fd8dbecf637eb5e67cdab5c5aa
01e0b7ac306895be84179f2715af269b
025dd300b0ccbd9d5898969c2a1ab138
02878fe3ccc81d4c884ca5574178d6a0
029cc71dc93341de90188b686798aa0d
02ef9a59d6da8b642271166d3ffd1b52
0305a4993ecf2d8ef4149fdfc7592603

(q2dev-amplicon) ➜  expected git:(gregs-review-pr81) ✗ biom table-ids --observations -i output_remove_contam_table/feature-table.biom | sort | head
003d45b24e8a58c0b20bf2bbfd7ff87f
008b8e3f9166408dc781796afd64a959
00dff50b45156cbb7debd5b5e82bf565
010b4e6583bfda219e2967b9b3973de0
010d323d97cc7ed0cd0081ec591d051a
012a37c5b111ebc368d03345b6049198
01d402b3bf484b51373700f1c1537b4b
025ebdf9e8907de3cb0010efa7424798
026a7407ef9739f4505fde62e29e6795
031b2591f8ff1aca82a3698fd490988c
```

you can see that the feature ids don't seem to align. 

So I think you thought this was an issue with your test, but it was actually an issue with your test data.

gregcaporaso committed:
update language in documentation

gregcaporaso committed:
various edits to simplify code and tests

1. simplified code and tests by removing all writing of files in favor of working with Python objects
2. removed dependence on artifacts in remove tests to better support diffs (even though the .biom files still aren't easily diff-able, b/c they're binary)
3. tried to improve variable names to make them more self-documenting
4. remove new skbio dependency (not done yet)
5. remove some unused functions

gregcaporaso reviewed (changes requested):
@jordenrabasco, I have a bunch of feedback on this one, and [created a PR](https://github.com/jordenrabasco/q2-quality-control/pull/1/) against your branch to clarify some of it as it is easier than trying to explain everything. I'm happy to chat (maybe Thursday?) to explain a bit if that's helpful. My PR won't work yet, but it's a step in the direction that I recommend moving with this. 

See my comments in this review. The biggest issue is that at the moment, your `remove` test isn't going to pass - I think I identified an issue with the test data that you're using. You'll see in my PR that I exported the artifacts that you're using in those tests, and am now using the data they contain directly (rather than the artifacts) - generally this is a good way to go as it makes for easier debugging since you can see the test data more easily, and generally speaking (though not in all cases) building your tests on data files or objects makes for more informative diffs if the test data changes across commits. 

I'm going to work on the test code in my PR a little more tomorrow. In the meantime, could you try to sort out the issue with the test data? We'll need that fixed to be able to move this forward. You can feel free to merge my PR (in which case those commits will show up in your PR - nothing will get merged to the main `dev` branch yet), or we can discuss and go from there.

jordenrabasco commented:
@gregcaporaso thanks so much for the going though all of that! 
That would be awesome, what time/day would work best for you to meet (Thursday I am free 10-noon, 3pm, and then after 6pm EST)?

Sorry about all the issues with the test data! Yes, prior to our meeting I will focus on investigating the issues with the testing data. I will also go through your comments here and in the PR on my branch, so we can hit the ground running during our meeting. Until then, I may hold off merging your PR into my branch just to avoid confusion on my end as I investigate the test data further.

gregcaporaso commented:
Thanks @jordenrabasco - I just sent you a calendar invite for Thursday with a Zoom link. Chat with you then!

gregcaporaso committed:
updated to use qiime2.plugin.util.transform in place of
self.transform_format

as the latter only looks for transformers in the current plugin.

tests are now working, but still not passing

gregcaporaso committed:
transition tests to use objects instead of files

gregcaporaso commented:
@jordenrabasco, I think I'm still waiting on you to let me know this is ready for a re-review. Is that right? Just want to make sure I didn't miss something.

jordenrabasco commented:
@gregcaporaso Ah my mistake, I thought I was waiting for you to upload some more changes to the PR on my fork before I pulled it in. I will pull in your PR on my fork and check to make sure everything is working by end of day today. Sorry for the confusion!

jordenrabasco committed:
Merge pull request #1 from gregcaporaso/gregs-review-pr81

changes to documentation and to simplify code and tests

jordenrabasco commented:
@gregcaporaso I just pulled in your PR into my fork and it looks like everything integrated correctly, so I would say that this PR is ready for re-review!

gregcaporaso commented:
Replaced by #85.
